geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/data/Repository.java

```diff
@@ -80,8 +88,95 @@ public Repository(OAuth2AuthorizedClientService authorizedClientService,
     this.clusterFactory = clusterFactory;
   }
 
+  /**
+   * this will return a cluster already connected to the geode jmx manager for the user in the
+   * request
+   * <p>
+   * But for multi-user connections to gemfireJMX, i.e pulse that uses gemfire integrated security,
+   * we will need to get the username from the context
+   */
+  public Cluster getCluster() {
+    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+    if (authentication == null) {
+      return null;
+    }
+
+    if (authentication instanceof OAuth2AuthenticationToken) {
+      return getClusterWithAuthenticationToken((OAuth2AuthenticationToken) authentication);
+    }
+
+    return getClusterWithUserNameAndPassword(authentication.getName(), null);
+  }
+
+  public Cluster getClusterWithUserNameAndPassword(String userName, String password) {
+    String[] credentials = {userName, password};
+    return getClusterWithCredentials(userName, credentials);
+  }
+
+  public Cluster getClusterWithCredentials(String userName, Object credentials) {
+    synchronized (clusterMap) {
+      Cluster cluster = clusterMap.get(userName);
+      if (cluster == null) {
+        logger.info(resourceBundle.getString("LOG_MSG_CREATE_NEW_THREAD") + " : " + userName);
+        cluster = clusterFactory.create(host, port, userName, resourceBundle, this);
+        // Assign name to thread created
+        cluster.setName(PulseConstants.APP_NAME + "-" + host + ":" + port + ":" + userName);
+        cluster.connectToGemFire(credentials);
+        if (cluster.isConnectedFlag()) {
+          clusterMap.put(userName, cluster);
+        }
+      }
+      return cluster;
+    }
+  }
+
+  /**
+   * Returns the cluster for the user associated with the given authentication. If the user's
+   * access token is expired, it is refreshed and the cluster is reconnected to JMX using the fresh
+   * token. If the refresh fails, the user's cluster is disconnected from JMX and removed from the
+   * repository.
+   */
+  private Cluster getClusterWithAuthenticationToken(OAuth2AuthenticationToken authentication) {
+    OAuth2AuthorizedClient authorizedClient = getAuthorizedClient(authentication);
+
+    if (isExpired(authorizedClient.getAccessToken())) {
```

Comment:
wondering which flow is better. Using the following might get rid of the reconnectToGemfire method in `Cluster`.
```suggestion
    String userName = authorizedClient.getPrincipalName();

    if (isExpired(authorizedClient.getAccessToken())) {
      logoutUser(userName);
      authorizedClient = refreshExpiredClient(authentication, authorizedClient);
    }

    userName = authorizedClient.getPrincipalName();
    String credentials = authorizedClient.getAccessToken().getTokenValue();
    return getClusterWithCredentials(userName, credentials);
```

Reply:
`logoutUser()` discards any data cached in the `Cluster`, including all of the trends stored in circular buffers. That seems like a harsh thing to do when the token refreshes.